src/profile-logic/marker-data.js

```diff
@@ -119,71 +119,192 @@ export function getSearchFilteredMarkerIndexes(
+  searchRegExps: MarkerRegExps,
+  categoryList: CategoryList
+): boolean {
   // Need to assign it to a constant variable so Flow doesn't complain about
   // passing it inside a function below.
-  const regExp = searchRegExp;
+  const regExps = searchRegExps;
+
   function test(value, key) {
+    key = key.toLowerCase();
+    const fieldRegexp = regExps.fieldMap.get(key);
+    const found =
+      (regExps.generic ? regExps.generic.test(value) : false) ||
+      (fieldRegexp && fieldRegexp.positive
+        ? fieldRegexp.positive.test(value)
+        : false);
+
     // Reset regexp for each iteration. Otherwise state from previous
     // iterations can cause matches to fail if the search is global or
     // sticky.
-    regExp.lastIndex = 0;
-
-    return (
-      regExp.test(value) ||
-      (!regExp.test(key + ':') && // avoid matching the key
-        // (eg. searching for 'me' should not match all markers,
-        //  even though they all have a 'name:' field).
-        regExp.test(key + ':' + value))
+    if (regExps.generic) {
+      regExps.generic.lastIndex = 0;
+    }
+    if (fieldRegexp && fieldRegexp.positive) {
+      fieldRegexp.positive.lastIndex = 0;
+    }
+    return found;
+  }
+
+  const { data, name, category } = marker;
+
+  if (categoryList[category] !== undefined) {
+    const markerCategory = categoryList[category].name;
+    if (test(markerCategory, 'cat')) {
+      return true;
+    }
+  }
+
+  if (test(name, 'name')) {
+    return true;
+  }
+
+  if (data && typeof data === 'object') {
+    if (test(data.type, 'type')) {
+      // Check the type of the marker payload first.
+      return true;
+    }
+
+    // Now check the schema for the marker payload for searchable
+    const markerSchema = getSchemaFromMarker(
+      markerSchemaByName,
+      marker.name,
+      marker.data
     );
+    if (
+      markerSchema &&
+      markerPayloadMatchesSearch(markerSchema, marker, test)
+    ) {
+      return true;
+    }
   }
 
-  const newMarkers: MarkerIndex[] = [];
-  for (const markerIndex of markerIndexes) {
-    const marker = getMarker(markerIndex);
-    const { data, name, category } = marker;
+  return false;
+}
 
-    if (categoryList[category] !== undefined) {
-      const markerCategory = categoryList[category].name;
-      if (test(markerCategory, 'cat')) {
-        newMarkers.push(markerIndex);
-        continue;
-      }
+function negativeFilterMarker(
+  marker: Marker,
+  markerSchemaByName: MarkerSchemaByName,
+  searchRegExps: MarkerRegExps,
+  categoryList: CategoryList
+): boolean {
+  // Need to assign it to a constant variable so Flow doesn't complain about
+  // passing it inside a function below.
+  const regExps = searchRegExps;
+
+  function test(value, key) {
+    key = key.toLowerCase();
+    const fieldRegexp = regExps.fieldMap.get(key);
+    const found =
+      fieldRegexp && fieldRegexp.negative
+        ? fieldRegexp.negative.test(value)
+        : false;
```

Comment:
I feel like that an early return if `fieldRegexp.negative` is absent would make it more readable.